Castling

bennypaulino opened the pull request:
This branch introduces [castling](https://en.wikipedia.org/wiki/Castling). This allows a player to move their king and rook only if certain requirements are met, as mentioned in the previously linked Wikipedia article:
- The king and the chosen rook are on the player's first rank.
- Neither the king nor the chosen rook has previously moved.
- There are no pieces between the king and the chosen rook.
- The king is not currently in check.
- The king does not pass through a square that is attacked by an enemy piece.
- The king does not end up in check. (True of any legal move.)

![castling](https://cloud.githubusercontent.com/assets/16200483/15982045/9dd16b6c-2f4c-11e6-9dc6-e6af52464992.gif)

This branch does the following:
- Adds a moved column to the pieces table, which helps determine whether a piece has moved during the course of a game.
- Modifies the parameters for update_attributes! for the move_to! method within the Piece model.
- Introduces valid_castling? to the King class's valid_move? method.
- Adds a King-class move_to! method that places the rook in its appropriate position to complete the castling move, if all of the aforementioned requirements are met.
- Includes 25 RSpec tests relevant to castling (found within app/spec/models/king_spec.rb).

bennypaulino committed:
Add a moved column to the pieces table.

This will indicate, either true or false, whether a piece has been moved or not

bennypaulino committed:
Call the obstructed? method for castling

bennypaulino committed:
Adjust the update_attributes! method within the Piece model

Include the moved flag for update_attributes!
Call the moved method on the king and rook.

bennypaulino committed:
Ensure that castling only occurs on the same rank

Limit castling to only the valid files (x_coords) relevant to castling.

bennypaulino committed:
Add comments for the methods involved with castling.

bennypaulino committed:
Add square_threatened? method

Use square_threatened? to help determine if castling is allowed.
The squares that the king travels through cannot be under threat
for castling to be permitted.

bennypaulino committed:
Add king_traversal_under_attack?

Castling is prevented temporarily:
if the square on which the king stands,
or the square which it must cross,
or the square which it is to occupy,
is under threat by one or more of the opponent's pieces.

bennypaulino committed:
Refactor, rename methods, and add comments.

bennypaulino committed:
Refactor guard statement, and add move_to! method for a castled rook.

bennypaulino committed:
Add move_to! method that determines if castling

is attempted, and moves the king & rook to their appropriate positions.
This method completes the castling move, if it is attempted.

- Rename `rook_requirements` method to `rook_requirements_met?`
- Refactor the guard within `rook_requirements_met?`
- Refactor the last statement that calls `obstructed?`

bennypaulino committed:
Remove statement regarding in_check from the valid_move?

bennypaulino committed:
Resolve merge conflict, and adjust the King model.

Modify `king_traversal_under_attack?` and call `square_threatened_by?`
method from the Game model. Allow access to the `square_threatened_by?`
method by removing it from private.
Conflicts:
	app/models/king.rb

bennypaulino committed:
Incorporate RSpec tests for the King model

Add king_spec.rb, and include tests for all of the
requirements that must be met for castling to be permitted.

cllns commented on app/models/king.rb:
```diff
@@ -1,12 +1,90 @@
+    return true if (-1..1).cover?(x_offset) && (-1..1).cover?(y_offset)
 
-    true
+    valid_castling?(new_x, new_y)
```
Would `(-1..1).cover?(x_offset) && (-1..1).cover?(y_offset) ||  valid_castling?(new_x, new_y)` work?
The reason I say so is that `return true if...` is a smell

cllns commented on app/models/king.rb:
```diff
@@ -1,12 +1,90 @@
+  # Determine if castling is allowed.
+  def valid_castling?(new_x, new_y)
+    # The right to castle has been lost if the king has moved.
+    return false if moved?
```
This is OK, with all these guard clauses, but I think you can leverage exceptions here. Basically you move all these things into assertion methods, so something like:

```
def assert_king_has_not_moved!
  raise KingHasMovedError if moved?
end
```

then this line would just be `assert_king_has_not moved!` here, and you'd have a few of those in a row, one for each.
of these guard clauses. 

In this case, you'd have to put the assertions in a `begin` / `rescue` block and have the rescue case return false.
Does that make sense?

(To be clear, I think it's OK as-is, this just makes things clearer)

cllns commented on app/models/king.rb:
```diff
@@ -1,12 +1,90 @@
+  def rook_requirements_met?(new_x, new_y)
+    # Set the file for the rook to 1 if the player's intent is to Castle queen-side,
+    # otherwise it will be set to eight. File is a chess term that's analogous to x_coord.
+    rook_file = new_x == 3 ? 1 : 8
```
I think you can remove stuff about `file` unless it's necessary. why not just replace`file` with `x_coord`?

cllns commented on app/models/king.rb:
```diff
@@ -1,12 +1,90 @@
+    king_file_array.each do |file|
+      return true if game.square_threatened_by?(opponent_color, file, new_y)
+    end
+    false
```
try `any` instead of `each` here?

bennypaulino replied:
Starting to lag over here, but I put some stuff into play and tightened up
the code a bit. Assertion methods and leveraging exceptions will get
tackled before the stand-up meeting tomorrow :.)
Thanks!!
Benny

On Sun, Jun 12, 2016 at 6:51 PM, Sean Collins <notifications@github.com>
wrote:

> In app/models/king.rb
> <https://github.com/letmecheck/app/pull/64#discussion_r66729215>:
>
> > +
> > +  # Castling is prevented temporarily:
> > +  # if the square on which the king stands, or the square which it must cross,
> > +  # or the square which it is to occupy, is attacked by one or more of the opponent's pieces.
> > +  def king_traversal_under_attack?(new_x, new_y)
> > +    # The argument for new_x will be either 3 or 7 if castling is attempted.
> > +    king_file_array = (new_x == 3) ? [3, 4, 5] : [5, 6, 7]
> > +
> > +    # square_threatened_by? takes color as one of its arguments.
> > +    opponent_color = (color == 'white') ? 'black' : 'white'
> > +
> > +    # Determine if any squares from the king's origin to his destination are under threat.
> > +    king_file_array.each do |file|
> > +      return true if game.square_threatened_by?(opponent_color, file, new_y)
> > +    end
> > +    false
>
> try any instead of each here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/letmecheck/app/pull/64/files/89a876db175ae2087abeae033ac409d435027599#r66729215>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APczI8EJ97mvnq1l7mOnHNOHvwKjQmUCks5qLI2KgaJpZM4IzcIt>
> .
>

bennypaulino replied:
I looked into the difference as far as performance is concerned between `any?` and `each`. I came across this interesting [article](http://twohlix.com/2014/11/ruby-arrayeach-vs-enumerableany/). Since we're dealing with such a tiny set, and not dealing with millions or billions of elements, I'll gladly use `any?` to make the code more readable!

bennypaulino committed:
Refactor and adjust for clarity

- refactor valid_move?
- refactor king_traversal_under_attack?
- rename variable within rook_requirements_met? for clarity.

bennypaulino committed:
Resolve merge conflicts

Merge conflicts resulted from merging en passant branch.
Keep changes from both branches, review schema, and resolve Piece model

Conflicts:
	app/models/piece.rb
	db/schema.rb

steve-mcclellan commented:
:shipit:
